packages/config-plugins/src/utils/BuildProperties.types.ts

```diff
@@ -0,0 +1,40 @@
+import type { ExpoConfig } from '@expo/config-types';
+
+/**
+ * Rule to transform from config to build properties
+ *
+ * @example
+ * ```
+ * {
+ *   propName: 'expo.jsEngine',
+ *   propValueGetter: (config) => config.android?.jsEngine ?? config.jsEngine ?? 'jsc',
+ * }
+ * ```
+ * Will lookup a value through the `propValueGetter` and update to `android/gradle.properties` / `ios/Podfile.properties.json`
+ * with the `expo.jsEngine` property name and the lookuped value.
+ *
+ */
+
+/**
+ * Source config can be either expo config or generic config
+ */
+export type BuildPropertiesConfig = ExpoConfig | Record<string, any>;
+
+export interface ConfigToPropertyRuleType<SourceConfigType extends BuildPropertiesConfig> {
+  // Property name in `android/gradle.properties` or `ios/Podfile.properties.json`
+  propName: string;
+
+  // Passing config and get the property value
+  propValueGetter: (config: SourceConfigType) => string | null | undefined;
+}
+
+/**
+ * ConfigPlugin extended type for `withBuildPodfileProps` or `withBuildGradleProps` which accepting SourceConfigType generics
+ */
+export type BuildPropertiesConfigPlugin = <SourceConfigType extends BuildPropertiesConfig>(
+  config: ExpoConfig,
+  props: {
+    configToPropertyRules: ConfigToPropertyRuleType<SourceConfigType>[];
+    sourceConfig?: SourceConfigType;
+  }
+) => ExpoConfig;
```

Comment:
Perhaps there needs to be a different name for this type of function since config plugins should, in theory, be fully serializable (i.e. props cannot be functions). This is more of a factory-type method like [this](https://github.com/expo/expo-cli/blob/139a09fbc04e601b0f3fe6d37ad1adedf27b1159/packages/config-plugins/src/plugins/ios-plugins.ts#L20) since it generates functionality and cannot be used in a fully JSON config.